osquery/tables/ipmi/ipmi_info.cpp

```diff
@@ -0,0 +1,1297 @@
+  rowsQueue_[key] = row;
+}
+
+void IPMIClient::toQueryData(QueryData& results) {
+  for (const auto& mcData : rowsQueue_) {
+    results.push_back(mcData.second);
+  }
+}
+
+void IPMIClient::pushParmData(parmData* parm) {
+  WriteLock lock(parmMutex_);
+  parmsOnHeap_.push_back(parm);
+}
+
+void IPMIClient::rmParmData(parmData* parm) {
+  WriteLock lock(parmMutex_);
+  // Check if parm exists first.
+  auto i = std::find(parmsOnHeap_.begin(), parmsOnHeap_.end(), parm);
+  if (i == parmsOnHeap_.end()) {
+    return;
+  }
+
+  std::unique_ptr<parmData, std::function<void(parmData*)>> _(
+      parm, [&](parmData* p) {
+        delete p;
+        parmsOnHeap_.erase(i);
+      });
+}
+
+void IPMIClient::blkAndOp(std::function<bool()> ready, int timeoutDurMS) {
+  std::atomic<bool> done(false);
+
+  auto fut = std::async(std::launch::async, [&, this] {
+    while (timeoutDurMS > 0 && !done.load()) {
+      pauseMilli(5);
+      if (interrupted()) {
+        return;
+      }
+
+      timeoutDurMS -= 6;
+    }
+  });
+  std::future_status status;
+
+  do {
+    int rv = oneOp();
+    if (rv == EINTR) {
+      done.store(true);
+      return;
+    }
+    if (rv != 0) {
+      LOG(ERROR) << "Could not handle IPMI event: " << strerror(rv);
+    }
+
+    status = fut.wait_for(std::chrono::milliseconds(1));
+
+  } while (status != std::future_status::ready && !ready());
+
+  done.store(true);
+}
+
+void IPMIClient::setLANCh(unsigned int ch) {
+  lanCh_ = ch;
+}
+
+void IPMIClient::findLANCh() {
+  iterateMCs([](ipmi_domain_t* domain, ipmi_mc_t* mc, void* data) {
+    for (unsigned int i = 0; i < 16; i++) {
+      int rv = ipmi_mc_channel_get_info(
+          mc,
+          i,
+          [](ipmi_mc_t* mc, int err, ipmi_channel_info_t* info, void* data) {
+            if (err != 0) {
+              TLOG << "Unexpected error on ipmi_mc_channel_get_info "
+                      "callback: "
+                   << strerror(err);
+              return;
+            }
+
+            unsigned int ch;
+            int rv = ipmi_channel_info_get_channel(info, &ch);
+            if (rv != 0) {
+              LOG(ERROR) << "Could not get channel: " << strerror(rv);
+              return;
+            }
+
+            unsigned int medium;
+            rv = ipmi_channel_info_get_medium(info, &medium);
+            if (rv != 0) {
+              LOG(ERROR) << "Could not get medium info for channel "
+                         << std::to_string(ch) << ": " << strerror(rv);
+              return;
+            }
+
+            if (medium == IPMI_CHANNEL_MEDIUM_8023_LAN) {
+              IPMIClient* c = reinterpret_cast<IPMIClient*>(data);
+              c->setLANCh(ch);
+            }
+
+          },
+          data);
+      if (rv != 0) {
+        LOG(ERROR) << "Could not get channel info for channel "
+                   << std::to_string(i) << ": " << strerror(rv);
+      }
+    }
+  });
+  blkAndOp([]() { return false; }, 500);
```

Comment:
This has been tuned from the nature of the queries, some take longer (perhaps more things to walk thru) than others.  So these numbers are somewhat heuristic.   I think it's absolutely worth cleaning up as preprocessor defined values.